### DISCUSSION TITLE
make a compound widget from existing elements?

### DISCUSSION THREAD

**Discussion by retsyo** [zauberzeug/nicegui#2188] [Q&A]
### Question

`ui.markdown` does not support `LaTeX` formula. We can use [MathJax](https://www.mathjax.org/) for this purpose

```python
ui.add_head_html(r'''
  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script id="MathJax-script" async
          src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js">
  </script>
''')

ui.html(r'''
<p>
 $$
 x = {-b \pm \sqrt{b^2-4ac} \over 2a}
 $$
</p>
''')
```

So a workaround to deal more general markdown string could be 

```python
def myMarkdown(strMd):
    lstMdSegment = fnSplitLatexOut(strMd)
    for strMdSeg in lstMdSegment:
        if strMdSeg.isLatex:
            ui.html(f'<p>{strMdSeg}</p>')
        else:
            ui.markdown(f'{strMdSeg}')
            
    return WHAT_SHOULD_WE_DO_HERE?     
    
```

but how to specify the return widget from this `myMarkdown`, so that it can be used as a normal NiceGUI Element/Control, I mean if  

```python
strMd = ```
The formula 
$$
 x = {-b \pm \sqrt{b^2-4ac} \over 2a}
$$
gives the roots described in following table:
| $$\sqrt{b^2-4ac}$$ | <0              | =0                 | >0                     |
| ------------------ | --------------- | ------------------ | ---------------------- |
|                    | 2 complex roots | 2 equal real roots | 2 different real roots |
```
then we can write
```python
with ui.button():
    ui.icon(something)
    myMarkdown(strMd)
```
thanks

**Comment by falkoschindler** [comment]
I'm not sure if this is what you're looking for, but you can create your own element like this:

```py
class my_element(ui.element):

    def __init__(self, text: str) -> None:
        super().__init__()
        if text.startswith('<'):
            ui.html(text)
        else:
            ui.label(text)

my_element('Text')
my_element('<b>HTML</b>')
```

In this example, `my_element` wraps a `ui.html` or a `ui.label` depending on `text`.

**Comment by retsyo** [comment]
so bad, if I use `ui.markdown` to show `table` or `code fence`, then `MathJax` does not render LaTeX formula

```python
from nicegui import ui
ui.add_head_html(r'''
  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script id="MathJax-script" async
          src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js">
  </script>
''')

ui.html(r'''
<p>
let \(\Delta = \sqrt{b^2-4ac}\), we can write

$$
 x = {-b \pm \sqrt {\Delta} \over 2a}
$$

We have a conclusion on it
</p>
''')

strTab = r'''| $$\Delta$$ | <0              | =0                 | >0                     |
| ------------------ | --------------- | ------------------ | ---------------------- |
|                    | 2 complex roots | 2 equal real roots | 2 different real roots |
'''
#~ ui.markdown(strTab)  # bad, this disables above MathJax

strCode = r'''
we can express this in code:
_```python    <---please delete the leading _ and this comment
delta = b**2-4*a*c

if delta<0:
    print('2 complex roots')
elif delta==0:
    print('2 equal real roots')
else:
    print('2 different real roots')
_```    <---please delete the leading _ and this comment
'''
#~ ui.markdown(strCode) # bad, this disables above MathJax

ui.run()
```

**Comment by stepheweffie** [comment]
> "To put mathematics in your web page, you can use TeX and LaTeX notation, MathML notation, AsciiMath notation, or a combination of all three within the same page;  " 
see docs - 
[https://docs.mathjax.org/en/latest/basic/mathematics.html](url)

The strCode is breaking the page due to the head script. It tolerates the str_Tab 

Here I'm passing your strings as strings in the page to the respective elements. 

<img width="1440" alt="Screen Shot 2024-01-01 at 7 18 42 PM" src="https://github.com/zauberzeug/nicegui/assets/1074503/9d4b2206-9794-4cf5-8eb2-ec2c557ddff3">

```
from nicegui import ui
from latex_demo import html, strTab, strCode


class LaTeXTemplate(ui.element):
    def __init__(self, script: str, template: str):
        super().__init__()
        self.script = script
        self.template = template
        ui.add_head_html(self.script)
        ui.html(self.template)

```

```
head_html = '''
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script type="text/javascript" id="MathJax-script" async
  src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js">
</script>
'''
```

```
body_html = html

@ui.page(path='/')
def index():
    with ui.row():
        LaTeXTemplate(head_html, body_html)
    with ui.row():
        ui.markdown(strTab)
    with ui.row():
        ui.markdown(strCode)

ui.run()
```

And passing a similar markdown with code, no comments breaks the MathJax - 
<img width="1440" alt="Screen Shot 2024-01-01 at 7 31 43 PM" src="https://github.com/zauberzeug/nicegui/assets/1074503/92122a48-496d-4d49-9ff7-284c5c7b8f71">

**Reply by stepheweffie** [reply]
This is what your strings should do in my code - 
<img width="1440" alt="Screen Shot 2024-01-01 at 7 11 25 PM" src="https://github.com/zauberzeug/nicegui/assets/1074503/9d64f063-78f7-4fa7-871e-cc7c5543e24a">
